File: interop/swift-enums/OldStyleViewController.m
//
//  OldStyleViewController.m
//  interop
//
//  Created by Daniel Wallace on 11/03/16.
//  Copyright © 2016 IkeGPS. All rights reserved.
//

#import "OldStyleViewController.h"
#import "interop-Swift.h"

@interface OldStyleViewController ()

@property (strong, nonatomic) UIColor *background;

@end

@implementation OldStyleViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    [self.view setBackgroundColor: [UIColor greenColor]];
    
    [self addBackgroundIcon];
    
    [self setTitle:@"Legacy Objective-C View Controller"];
    
    NSLog(@"\nIn Objective-C, Icon enum is %@", self._icon);
    
    NSLog(@"self.iconImage = %@", self.iconImage);
    NSLog(@"iconColor = %@", self.iconColor);

    NSLog(@"Pause");
}

- (void)addBackgroundIcon {
    if (self.background && self.view) {
        UIView *newView = [[UIView alloc] initWithFrame:[self iconFrame]];
        [newView setBackgroundColor:self.background];
        [self.view addSubview:newView];
    }
}

-(CGRect)iconFrame {
    return CGRectMake(0, 0, self.view.bounds.size.width/2, self.view.bounds.size.height/2);
}

/**
 implementation of shim to allow passing a Swift enum to an Objective-C class
 see OldStyleViewController.swift
 */
-(void)showIconViewWithColor:(UIColor *)color{
    self.background = color;
    [self addBackgroundIcon];
}

/**
 implementation of shim to allow passing a Swift enum to an Objective-C class
 see OldStyleViewController.swift
 */
-(void)showIconViewWithImage:(UIImage *)image {
    UIColor *backgroundImage = [[UIColor alloc] initWithPatternImage:image];
    self.background = backgroundImage;
    [self addBackgroundIcon];
}


@end

File: interop/swift-enums/OldStyleViewController.swift
//
//  LegacyViewController.swift
//  interop
//
//  Created by Daniel Wallace on 11/03/16.
//  Copyright © 2016 IkeGPS. All rights reserved.
//


extension OldStyleViewController {
    
    // Can store the Icon enum and pass it on to something else. Not usable in Obj-C code though.
    var icon: Icon {
        get {
            return (_icon as! Box<Icon>).value
        }
        set{
            _icon = Box(newValue)
        }
    }
    
    @objc func addIconView() {
        
        let iconView = IconView(icon: icon)
        self.view.addSubview(iconView)
        
        let views:[String: AnyObject] = ["topLayoutGuide": topLayoutGuide,"bottomLayoutGuide": bottomLayoutGuide, "view": view, "iconView": iconView]
        
        var allConstraints = [NSLayoutConstraint]()
        
        let verticalConstraints = NSLayoutConstraint.constraintsWithVisualFormat(
            "V:[iconView]-(50)-|",
            options: [],
            metrics: nil,
            views: views)
        allConstraints += verticalConstraints
        
        let horizontalConstraints = NSLayoutConstraint.constraintsWithVisualFormat(
            "H:|-[iconView]",
            options: [],
            metrics: nil,
            views: views)
        allConstraints += horizontalConstraints
        
        NSLayoutConstraint.activateConstraints(allConstraints)
    }
    
    /**
     a shim to allow setting an Objective-C property from a Swift enum (Icon.Color)
     */
    @objc var iconColor: UIColor? {
        get {
            switch icon {
            case let .Color(color) : return color
            case .Image : return nil
            }
        }
        set {
            guard newValue != nil else { return }
            icon = .Color(newValue!)
        }
    }
    
    /**
     a shim to allow setting an Objective-C property from a Swift enum (Icon.Image)
     */
    @objc var iconImage: UIImage? {
        get {
            switch icon {
            case let .Image(image) : return image
            case .Color : return nil
            }
        }
        set {
            guard newValue != nil else { return }
            icon = .Image(newValue!)
        }
    }
    
    /**
     a shim to allow passing a Swift enum to an Objective-C class
     */
    func showIconViewWithIcon(icon: Icon) {
        switch icon {
        case let .Image(image):
            self.showIconViewWithImage(image)
        case let .Color(color):
            self.showIconViewWithColor(color)
        }
    }
}


